Added glob matching and ignore files

lukeburns opened the pull request:
I wanted a .datignore file for dat that worked like .gitignore for git. I decided to implement this here in folder-walker rather than just use the filter option so that I didn't have to walk the tree twice to find ignore files. It uses minimatch, the library used internally by npm. 

It's heavier, but allows you to replace

```js
var stream = walker(dirs, {filter: function (data) {
    if (path.basename(data) === '.dat') return false
    return true
}})
```

with

```js
var stream = walker('/path/to/folder', { ignore: ['.dat'], ignoreFiles: ['.datignore'] })
```

and get ignore files along for the ride. 

If you're on board with this, I have a pull request pretty much lined up that adds .datignore support to dat.

mafintosh commented:
The `ignore` option looks okay to me. I'm not 100% sure about the `ignoreFiles` option as that seems out of scope for this module since it forces users to use a specific format for ignore files.

Btw minimatch doesn't work like .gitignore - we learned that the hard way when adding ignore file support to https://github.com/feross/standard. Use the ignore module for that instead, https://github.com/kaelzhang/node-ignore#readme

lukeburns commented:
I see that. The benefit to implementing ignore files here is that it reduces the number of times needed to traverse the directory structure to find the ignore files. How important is this for dat?

Thanks for pointing to node-ignore.

mafintosh commented:
Good point about only traversing once...

lukeburns commented:
This handles ignore files in multiple directories incorrectly. It currently adds rules from each ignore file it comes to a global set of rules. It should have a set of rules specific to directory it's working through – rules given by parents but not siblings or children.

karissa commented:
Ah this makes sense. Luke, thanks for taking the initiative here! I am a little weary about adding the `ignoreFiles` option too. 

In general, I'd like to keep this module as simple as possible to reduce the possibility of bugs. I like having only a `filter` function implemented by the developer because it can be very flexible and its clear what it does.

karissa commented:
That having been said, I think `ignore` is a common enough operation that we could include it here as an option alongside `filter` and I'd be happy merging that.

lukeburns commented:
I've removed ignore file support and added a `hook` option which allows you to filter directory by directory – i.e. to identify ignore files in a directory, extract ignore rules, apply those rules, and pass filtered paths back into `pending`. I've added an example file which includes a possible ignorefile implementation (which handles ignore files in multiple directories correctly now).

```js
var stream = walker({ hook: function (directory, files, next) {
   var paths = filter(directory, files) // filter files based on ignore files in directory
   next(paths) // pass filtered paths to pending
}})
```

karissa commented:
Oh wow, cool! Thanks for this. I think I like the hook interface and it is a powerful abstraction, although it might be a little bit harder to understand at first.

lukeburns commented:
The example file makes it pretty easy to catch on to what's happening, but I agree, it's not quite as straightforward as it could be.

The other option I considered was for the hook to add local directory ignore rules, which is more limiting then this abstraction, but much more straightforward. It would be used like

'''js
var stream = walkers({hook: function(directory, files, apply) { 
   var localRules = getRules(directory, files) // array of ignore rules 
   apply(localRules) // applies to directory and its subdirectories 
'''
